source/UF/Pushouts.lagda

```diff
@@ -0,0 +1,418 @@
+ : {A : 𝓤  ̇} {B : 𝓥  ̇} {C : 𝓦  ̇} (S : 𝓤'  ̇)
+   (f : C → A) (g : C → B) (s : cocone f g S) (P : S →  𝓣  ̇)
+ → ((x : S) → P x) → dependent-cocone f g S s P
+canonical-map-to-dependent-cocone S f g (i , j , G) P d =
+ (d ∘ i , d ∘ j , λ c → apd d (G c))
+
+Pushout-Dependent-Universal-Property
+ : {A : 𝓤  ̇} {B : 𝓥  ̇} {C : 𝓦  ̇} (S : 𝓤'  ̇) 
+   (f : C → A) (g : C → B) (s : cocone f g S) (P : S →  𝓣  ̇)
+ → 𝓤 ⊔ 𝓥 ⊔ 𝓦 ⊔ 𝓤' ⊔ 𝓣  ̇
+Pushout-Dependent-Universal-Property S f g s P =
+ is-equiv (canonical-map-to-dependent-cocone S f g s P)
+
+Pushout-Induction-Principle
+ : {A : 𝓤  ̇} {B : 𝓥  ̇} {C : 𝓦  ̇} (S : 𝓤'  ̇)
+   (f : C → A) (g : C → B) (s : cocone f g S) (P : S → 𝓣  ̇)
+ → 𝓤 ⊔ 𝓥 ⊔ 𝓦 ⊔ 𝓤' ⊔ 𝓣  ̇
+Pushout-Induction-Principle {_} {_} {_} {_} {_} {A} {B} {C} S f g (i , j , G) P 
+ = (l : (a : A) → P (i a))
+ → (r : (b : B) → P (j b))
+ → ((c : C) → transport P (G c) (l (f c)) ＝ r (g c))
+ → (x : S) → P x
+
+Pushout-Computation-Rule₁
+ : {A : 𝓤  ̇} {B : 𝓥  ̇} {C : 𝓦  ̇} (S : 𝓤'  ̇) 
+   (f : C → A) (g : C → B) (s : cocone f g S) (P : S → 𝓣  ̇)
+ → Pushout-Induction-Principle S f g s P
+ → 𝓤 ⊔ 𝓥 ⊔ 𝓦 ⊔ 𝓣  ̇
+Pushout-Computation-Rule₁
+ {_} {_} {_} {_} {_} {A} {B} {C} S f g (i , j , G) P S-ind
+ = (l : (a : A) → P (i a))
+ → (r : (b : B) → P (j b))
+ → (H : (c : C) → transport P (G c) (l (f c)) ＝ r (g c))
+ → (a : A)
+ → S-ind l r H (i a) ＝ l a
+
+Pushout-Computation-Rule₂
+ : {A : 𝓤  ̇} {B : 𝓥  ̇} {C : 𝓦  ̇} (S : 𝓤'  ̇)
+   (f : C → A) (g : C → B) (s : cocone f g S) (P : S → 𝓣  ̇)
+ → Pushout-Induction-Principle S f g s P
+ → 𝓤 ⊔ 𝓥 ⊔ 𝓦 ⊔ 𝓣  ̇
+Pushout-Computation-Rule₂
+ {_} {_} {_} {_} {_} {A} {B} {C} S f g (i , j , G) P S-ind
+ = (l : (a : A) → P (i a))
+ → (r : (b : B) → P (j b))
+ → (H : (c : C) → transport P (G c) (l (f c)) ＝ r (g c))
+ → (b : B)
+ → S-ind l r H (j b) ＝ r b
+
+Pushout-Computation-Rule₃
+ : {A : 𝓤  ̇} {B : 𝓥  ̇} {C : 𝓦  ̇} (S : 𝓤'  ̇) 
+   (f : C → A) (g : C → B) (s : cocone f g S) (P : S → 𝓣  ̇)
+   (S-ind : Pushout-Induction-Principle S f g s P)
+ → Pushout-Computation-Rule₁ S f g s P S-ind
+ → Pushout-Computation-Rule₂ S f g s P S-ind
+ → 𝓤 ⊔ 𝓥 ⊔ 𝓦 ⊔ 𝓣  ̇
+Pushout-Computation-Rule₃
+ {_} {_} {_} {_} {_}{A} {B} {C} S f g (i , j , G) P S-ind S-comp₁ S-comp₂
+ = (l : (a : A) → P (i a))
+ → (r : (b : B) → P (j b))
+ → (H : (c : C) → transport P (G c) (l (f c)) ＝ r (g c))
+ → (c : C)
+ → apd (S-ind l r H) (G c) ∙ S-comp₂ l r H (g c)
+ ＝ ap (transport P (G c)) (S-comp₁ l r H (f c)) ∙ H c
+
+\end{code}
+
+The following are logically equivalent (which is an instance of Sojakova's result):
+
+1) The dependent universal property
+2) The induction principle with propositional computation rules
+3) The recursion principle with propositional computation rules and the
+   uniqueness principle
+4) The universal property.
+
+Below we will derive 2), 3) and 4) from the seemingly strongest assumption 1).
+In a later file we will attempt to derive 1), 2) and 3) from the seemingly weakest
+assumption 4) (this is a work in progress.)
```

Comment:
Probably remove article "a".

Reply:
I'm just going to reword this completly. Let me know if you have thoughts on the new wording.